src/graphics/webgl/webgl-shader.js

```diff
@@ -0,0 +1,333 @@
+     * @param {GraphicsDevice} device - The graphics device.
+     * @param {Shader} shader - The shader to restore.
+     * @ignore
+     */
+    restoreContext(device, shader) {
+        this.compileAndLink(device, shader);
+    }
+
+    /**
+     * Compile and link a shader program.
+     *
+     * @param {GraphicsDevice} device - The graphics device.
+     * @param {Shader} shader - The shader to compile.
+     * @ignore
+     */
+    compileAndLink(device, shader) {
+        const definition = shader.definition;
+        Debug.assert(definition.vshader, 'No vertex shader has been specified when creating a shader.');
+        Debug.assert(definition.fshader, 'No fragment shader has been specified when creating a shader.');
+
+        const glVertexShader = this.compileShaderSource(device, definition.vshader, true);
+        const glFragmentShader = this.compileShaderSource(device, definition.fshader, false);
+
+        const gl = device.gl;
+        const glProgram = gl.createProgram();
+
+        gl.attachShader(glProgram, glVertexShader);
+        gl.attachShader(glProgram, glFragmentShader);
+
+        const attrs = definition.attributes;
+        if (device.webgl2 && definition.useTransformFeedback) {
+            // Collect all "out_" attributes and use them for output
+            const outNames = [];
+            for (const attr in attrs) {
+                if (attrs.hasOwnProperty(attr)) {
+                    outNames.push("out_" + attr);
+                }
+            }
+            gl.transformFeedbackVaryings(glProgram, outNames, gl.INTERLEAVED_ATTRIBS);
+        }
+
+        // map all vertex input attributes to fixed locations
+        const locations = {};
+        for (const attr in attrs) {
+            if (attrs.hasOwnProperty(attr)) {
+                const semantic = attrs[attr];
+                const loc = semanticToLocation[semantic];
+                Debug.assert(!locations.hasOwnProperty(loc), `WARNING: Two attribues are mapped to the same location in a shader: ${locations[loc]} and ${attr}`);
+
+                locations[loc] = attr;
+                gl.bindAttribLocation(glProgram, loc, attr);
+            }
+        }
+
+        gl.linkProgram(glProgram);
+
+        // Cache the WebGL objects on the shader
+        this.glVertexShader = glVertexShader;
+        this.glFragmentShader = glFragmentShader;
+        this.glProgram = glProgram;
+
+        // #if _PROFILER
+        device._shaderStats.linked++;
+        if (definition.tag === SHADERTAG_MATERIAL) {
+            device._shaderStats.materialShaders++;
+        }
+        // #endif
+    }
+
+    /**
+     * Compiles an individual shader.
+     *
+     * @param {string} src - The shader source code.
+     * @param {boolean} isVertexShader - True if the shader is a vertex shader, false if it is a
+     * fragment shader.
+     * @returns {WebGLShader} The compiled shader.
+     * @private
+     */
+    compileShaderSource(device, src, isVertexShader) {
```

Comment:
If this is genuinely private, maybe prepend it with an undrescore?

Reply:
done